include/boost/geometry/algorithms/detail/closest_points/multipoint_to_geometry.hpp

```diff
@@ -0,0 +1,326 @@
+{
+    template 
+    <
+        typename MultiPoint1, 
+        typename MultiPoint2, 
+        typename Segment, 
+        typename Strategies
+    >
+    static inline void apply(MultiPoint1 const& multipoint1,
+                             MultiPoint2 const& multipoint2,
+                             Segment& shortest_seg,
+                             Strategies const& strategies)
+    {
+        if (boost::size(multipoint1) < boost::size(multipoint2))
+        {
+            point_or_segment_range_to_geometry_rtree::apply(
+                boost::begin(multipoint2),
+                boost::end(multipoint2),
+                multipoint1,
+                shortest_seg,
+                strategies);
+            detail::closest_points::swap_segment_points::apply(shortest_seg);
+            return;
+        }
+        point_or_segment_range_to_geometry_rtree::apply(
+            boost::begin(multipoint1),
+            boost::end(multipoint1),
+            multipoint2,
+            shortest_seg,
+            strategies);
+    }
+};
+
+struct multipoint_to_linear
+{
+    template 
+    <
+        typename MultiPoint, 
+        typename Linear, 
+        typename Segment, 
+        typename Strategies
+    >
+    static inline void apply(MultiPoint const& multipoint,
+                             Linear const& linear,
+                             Segment& shortest_seg,
+                             Strategies const& strategies)
+    {
+        point_or_segment_range_to_geometry_rtree::apply(
+            boost::begin(multipoint),
+            boost::end(multipoint),
+            linear,
+            shortest_seg,
+            strategies);
+    }
+};
+
+struct linear_to_multipoint
+{
+    template 
+    <
+        typename Linear, 
+        typename MultiPoint, 
+        typename Segment, 
+        typename Strategies
+    >
+    static inline void apply(Linear const& linear,
+                             MultiPoint const& multipoint,
+                             Segment& shortest_seg,
+                             Strategies const& strategies)
+    {
+        multipoint_to_linear::apply(multipoint, linear, shortest_seg, strategies);
+        detail::closest_points::swap_segment_points::apply(shortest_seg);
+    }
+};
+
+struct segment_to_multipoint
+{
+    template 
+    <
+        typename Segment, 
+        typename MultiPoint, 
+        typename OutSegment, 
+        typename Strategies
+    >
+    static inline void apply(Segment const& segment,
+                             MultiPoint const& multipoint,
+                             OutSegment& shortest_seg,
+                             Strategies const& strategies)
+    {
+        using linestring_type = geometry::model::linestring
+            <
+                typename point_type<Segment>::type
+            >;
+        linestring_type linestring;
+        convert(segment, linestring);
+        multipoint_to_linear::apply(multipoint, linestring, shortest_seg, strategies);
```

Comment:
Converting segment to linestring seems to be unnecessary but if it simplifies the implementation then I'm fine with it.